### DISCUSSION TITLE
Astro sitemap allow opting out of dashes in sitemap.xml

### DISCUSSION THREAD

**Discussion by Hacksore** [withastro/roadmap#940] [Proposal]
# Summary

The following places there is a hardcoded dash `-` so we should allow a way to opt-out of that for the SEO best practices.

https://github.com/withastro/astro/blob/ff8004f6a7b2aab4c6ac367f13744a341c3c5462/packages/integrations/sitemap/src/index.ts#L47
https://github.com/withastro/astro/blob/ff8004f6a7b2aab4c6ac367f13744a341c3c5462/packages/integrations/sitemap/src/write-sitemap.ts#L42

# Background & Motivation

Someone told me so it's bad practice to have hyphens/dashes and it can yield a site map to no load.
https://www.seroundtable.com/google-sitemaps-error-with-hyphenated-file-names-36829.html#:~:text=Google%20might%20be%20having%20some,will%20not%20trigger%20any%20errors.

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- SEO best practices

# Example

pseudo codeish for how It could work?

```ts
import sitemap from "@astrojs/sitemap";

export default defineConfig({
  integrations: [
    sitemap({
      // NOTE: new support for no dashes
      // omit the extension, it will always be .xml
      baseFilename: "sitemap",
      // base name, we omit .xml as they always have to be that
      filename: async ({ index, baseFilename }}) => {
        return baseFilename + index; // sitemap1.xml, sitemap2.xml, etc
      },
      // // NOTE: backwards compat which we can keep as default so no breaking changes?
      // // omit the extension, it will always be .xml
      // baseFilename: "sitemap-index",
      // // base name, we omit .xml as they always have to be that
      // filename: async ({ index, baseFilename }}) => {
      //   return `sitemap-${index}`; // this will be sitemap-1.xml, sitemap-2.xml, etc
      // },
    }),
  ],
});
```

**Comment by RhysSullivan** [comment]
Here's an example:

![image](https://github.com/withastro/roadmap/assets/39114868/3eae966e-ae0e-4182-b478-1bb19fd1130a)

**Comment by florian-lefebvre** [comment]
I think it could make sense to have a `separator` property defaulting to `'-'` in v4. Could make sense to default it to `'_'` in v5 if it's better for SEO:
```ts
export default defineConfig({
	integrations: [sitemap({
		separator: '_' // or ''
	})]
})
```

**Comment by adaliszk** [comment]
Why not provide a way to specify the entire pattern or define a pure function that generates the filename?
```typescript
export default defineConfig({
    integrations: [
        sitemap({
            filename: "sitemap{index}.xml",
        }),
    ],
});
```
or
```typescript
export default defineConfig({
    integrations: [
        sitemap({
            filename({ index }) {
                return `sitemap${index}.xml`;
            }, 
        }),
    ],
});
```
In both cases, the integration can expect a type of `${string}.xml` as an output, so people would be forced to provide a valid filename.

**Comment by Hacksore** [comment]
@adaliszk Should we allow the main sitemap file to be named differently than the index generated files?

**base file**: `sitemap-index.xml` 
**index files**: `sitemap-0.xml` ... `sitemap-13.xml`

**Reply by adaliszk** [reply]
well, there are also other patterns that should be supported, like https://developers.google.com/search/docs/crawling-indexing/sitemaps/build-sitemap#cross-submit: `https://www.example.com/sub1-example-com-sitemap.xml`

**Comment by Hacksore** [comment]
So this is what I am thinking now which should cover all use-cases. Let me know your thoughts @adaliszk @florian-lefebvre.

```ts
import sitemap from "@astrojs/sitemap";

export default defineConfig({
  integrations: [
    sitemap({
      // NOTE: new support for no dashes
      // omit the extension, it will always be .xml
      baseFilename: "sitemap",
      // base name, we omit .xml as they always have to be that
      filename: async ({ index, baseFilename }}) => {
        return baseFilename + index; // sitemap1.xml, sitemap2.xml, etc
      },
      // // NOTE: backwards compat which we can keep as default so no breaking changes?
      // // omit the extension, it will always be .xml
      // baseFilename: "sitemap-index",
      // // base name, we omit .xml as they always have to be that
      // filename: async ({ index, baseFilename }}) => {
      //   return `sitemap-${index}`; // this will be sitemap-1.xml, sitemap-2.xml, etc
      // },
    }),
  ],
});
```

**Reply by adaliszk** [reply]
I quite like having hooks so that you can pull in any logic to resolve the name you'd like. One thing that should be supported is async-await just in case some logic is promise-based.

**Reply by Hacksore** [reply]
@adaliszk SGTM, I've updated the original post as well as this one.

**Comment by Hacksore** [comment]
Another thing that might be relevant here is how to handle two domains and one static app.

ie. `x.app` and `x.com`

@RhysSullivan do you know what’s the typical way to do this?

my guess is dynamic robots.txt and sitemap 😳.

**Comment by vin-e** [comment]
Good work on this thus far @Hacksore.

I am currently rebuilding a site from Eleventy to Astro. The existing site uses `sitemap.xml` as the base file. I would appreciate a way to customize the name or remove the `-index` from the file.  I currently do not have access to the google search console to make the appropriate updates. 

Thanks

**Reply by Hacksore** [reply]
yeah then I think we'd want something like this before we give a better programitic api?

```ts
sitemap({
  separator: "", // default: -
  indexFilename: "sitemap" // default: sitemap-index
})
```

If that seems agreeable I can update my PR https://github.com/withastro/astro/pull/11898
